### PR TITLE
Calling HeadObjectWithContext() concurrently to check the existence of a module

### PR DESCRIPTION
**Pull request by linzhp**
This is an alternative implementation of #1840, with goroutines.

@manugupt1 @ngshiheng @r-ashish

### CODE REVIEW

File: pkg/storage/s3/checker.go [hunk 1]
```diff
@@ -19,27 +19,31 @@ func (s *Storage) Exists(ctx context.Context, module, version string) (bool, err
...
+	cancelingCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, file := range files {
+		go func(file string) {
+			_, err := s.s3API.HeadObjectWithContext(
+				cancelingCtx,
+				&s3.HeadObjectInput{
+					Bucket: aws.String(s.bucket),
+					Key:    aws.String(config.PackageVersionedName(module, version, file)),
+				})
```
Comment (marwan-at-work) [inline pkg/storage/s3/checker.go:34]: is that the implementation of the `Exists()` method, if so can we just call `.Exists()` instead? And if not, any reason not to make the Exists() method use this logic?

**Review by marwan-at-work — APPROVED** [review]
Nevermind my previous comment 👍🏼 PS. Not an AWS/S3 expert but the concurrency part looks good to me

File: pkg/storage/s3/checker.go [hunk 2]
```diff
@@ -19,27 +19,31 @@ func (s *Storage) Exists(ctx context.Context, module, version string) (bool, err
...
-			}
-			if *o.Key == config.PackageVersionedName(module, version, "info") ||
-				*o.Key == config.PackageVersionedName(module, version, "mod") ||
-				*o.Key == config.PackageVersionedName(module, version, "zip") {
-				found[*o.Key] = struct{}{}
+	for range files {
+		err := <-errChan
+		if err != nil {
+			var aerr awserr.Error
+			if errs.As(err, &aerr) && aerr.Code() == "NotFound" {
```
Comment (marwan-at-work) [inline pkg/storage/s3/checker.go:42]: As a side note, we should probably create a `function As` inside our `errors` package that just proxies to the standard library one, this way you don't need to import and rename to `errs` here.

File: pkg/storage/s3/checker.go [hunk 3]
```diff
@@ -19,27 +19,31 @@ func (s *Storage) Exists(ctx context.Context, module, version string) (bool, err
 	ctx, span := observ.StartSpan(ctx, op.String())
 	defer span.End()
 
-	lsParams := &s3.ListObjectsInput{
-		Bucket: aws.String(s.bucket),
-		Prefix: aws.String(fmt.Sprintf("%s/@v", module)),
+	files := []string{"info", "mod", "zip"}
+	errChan := make(chan error, len(files))
+	defer close(errChan)
+	cancelingCtx, cancel := context.WithCancel(ctx)
```
Comment (manugupt1) [inline pkg/storage/s3/checker.go:26]: I think this should go in the go routine so that each remote cal has its own ctx

**Review by manugupt1 — APPROVED** [review]

File: pkg/storage/s3/checker.go (hunk 3 above)
Comment (linzhp) [reply pkg/storage/s3/checker.go:26]: Having all goroutine sharing the same `cancelingCtx` allows one failing goroutine to cancel other calls to S3: if we already know one of the file don't exist, we can cancel other `HeadObject` calls.

**Comment by linzhp** [comment]
All unit tests passed. Deployed to Uber's internal Go proxy and verified it worked.

File: pkg/storage/s3/checker.go (hunk 2 above)
Comment (nrwiersma) [reply pkg/storage/s3/checker.go]: A proxy function already exists: `errors.AsErr`.

**Pull request merged**